File: cs/src/ConstraintSolver/CNSAT/Decider.cs
using System;
using System.Collections.Generic;

namespace Alica.Reasoner.CNSAT
{
	public class Decider
	{
		public Decider()
		{
		}
		public static Var DecideRangeBased(List<Var> variables, CNSat solver) {
			List<Lit> choices = new List<Lit>();
			int vars = variables.Count;
			for(int i=0; i<vars; i++) {				
				if (variables[i].Assignment != Assignment.Unassigned) continue;
				bool hT = false;
				bool hF = false;
				for(int j=0; j<variables[i].WatchList.Count; j++) {					
					if(!variables[i].WatchList[j].Clause.Satisfied) {						
						if (variables[i].WatchList[j].Lit.Sign == Assignment.True) {
							if(!hT) choices.Add(variables[i].WatchList[j].Lit);	
							hT = true;
							
						} else {
							if(!hF) choices.Add(variables[i].WatchList[j].Lit);
							hF = true;							
						}						
					}
					if (hT && hF) break;
				}
			}
			if(choices.Count == 0) return null;
			
			choices.Sort(LitRangeCompare);
			/*Console.WriteLine("======");
			foreach(Lit l in choices) {
				Console.WriteLine("{0}",(l.Sign == Assignment.True ? l.Var.PositiveRangeSize:l.Var.NegativeRangeSize));
			}
			Console.WriteLine("======");*/
			//Choices are now in ascending order!
			Var v;
			Assignment ass;
			
			//Uniform:
			//int idx = solver.Rand.Next(choices.Count);
			//smallest range:
			//int idx = 0;
			//largest range:
			int idx = choices.Count-1;
			//Prefer smaller:
			/*int idx = choices.Count-1;
			double r = solver.Rand.NextDouble();
			double q = 0;
			for(int i=0; i<choices.Count; i++) {
				q += 1.0/(i+2);
				if (q > r) {
					idx = i;
					break;
				}
			}*/
			//Prefer larger:
			/*int idx = 0;
			double r = solver.Rand.NextDouble();
			double q = 0;
			for(int i=0; i<choices.Count; i++) {
				q += 1.0/(i+2);
				if (q > r) {
					idx = choices.Count-i-1;
					break;
				}
			}*/
			//Pick middle:
			//int idx = choices.Count/2;
			
			v = choices[idx].Var;
			ass = choices[idx].Sign;
			
			DecisionLevel d = new DecisionLevel(solver.Decisions.Count);
			solver.DecisionLevel.Add(d);					
			v.Assignment = ass;
			solver.Decisions.Add(v);
			v.Reason=null;
			v.DecisionLevel = d;
			
			return v;
			
		}
		
		
		public static Var DecideActivityBased(List<Var> variables, CNSat solver) {
			int vars = variables.Count;
			Random r = solver.r;
			int init = r.Next(vars);
			Var v=null, next=null;
			int maxActivity=0;
			//Search Lit with highest Activity
			if(solver.CNSMTGSolver==null) {
				for(int i=0; i<vars; i++) {
					int p = (init+i)%vars;
					if(p<0||p>=vars) Console.WriteLine("p = " +p);
					v = variables[p];
					if(v.Assignment == Assignment.Unassigned) {
						if(maxActivity<=v.Activity) {
							maxActivity=v.Activity;
							next=v;
						}
					}
				}
				//Decide it
				if(next!=null) {
					DecisionLevel d = new DecisionLevel(solver.Decisions.Count);
					solver.DecisionLevel.Add(d);
					
					double rel = next.PositiveAppearance+next.NegativeAppearance;
					if(rel!=0) rel = ((double)next.PositiveAppearance)/rel;
					else rel = 0.5;
					
					next.Assignment = (r.NextDouble() < rel) ? Assignment.True : Assignment.False;
					solver.Decisions.Add(next);
					next.Reason=null;
					next.DecisionLevel = d;
					return next;
				}
			} else { 
				init = r.Next(solver.Clauses.Count);
				for(int i=0; i<solver.Clauses.Count; i++) {
					Clause c = solver.Clauses[(i+init)%solver.Clauses.Count];
					if(!c.Satisfied && c.Literals.Count>1) {
						if(!c.watcher[0].Lit.Satisfied()) {
							next = c.watcher[0].Lit.Var;
							DecisionLevel d = new DecisionLevel(solver.Decisions.Count);
							solver.DecisionLevel.Add(d);
							
							next.Assignment = c.watcher[0].Lit.Sign;
							solver.Decisions.Add(next);
							next.Reason=null;
							next.DecisionLevel = d;
							return next;
						}
						else if(!c.watcher[1].Lit.Satisfied()) {
							DecisionLevel d = new DecisionLevel(solver.Decisions.Count);
							solver.DecisionLevel.Add(d);
							next = c.watcher[1].Lit.Var;
							
							next.Assignment = c.watcher[1].Lit.Sign;
							solver.Decisions.Add(next);
							next.Reason=null;
							next.DecisionLevel = d;
							return next;
						}
						else Console.WriteLine("This shoud Never Happen!!");
					}
				}
			}
			
			return null;
		}
		
		public static Var DecideVariableCountBased(List<Var> variables, CNSat solver) {
			int vars = variables.Count;
			Lit l = null;
			int maxCount = Int32.MaxValue;
			int minCount = -1;
			
			for(int i=0; i<vars; i++) {				
				if (variables[i].Assignment != Assignment.Unassigned) continue;
				for(int j=0; j<variables[i].WatchList.Count; j++) {
					if(!variables[i].WatchList[j].Clause.Satisfied) {
						/*if(maxCount>variables[i].WatchList[j].Lit.VariableCount) {
							l = variables[i].WatchList[j].Lit;
							maxCount = l.VariableCount;
						}*/
						if(minCount<variables[i].WatchList[j].Lit.VariableCount) {
							l = variables[i].WatchList[j].Lit;
							minCount = l.VariableCount;
						}
					}
				}
			}
			if(l==null) return null;
			Assignment ass;
			Var v=l.Var;
					
			ass = l.Sign;
			
			DecisionLevel d = new DecisionLevel(solver.Decisions.Count);
			solver.DecisionLevel.Add(d);					
			v.Assignment = ass;
			solver.Decisions.Add(v);
			v.Reason=null;
			v.DecisionLevel = d;
			
			return v;
			
		}
		
		
		
		
		public static int LitRangeCompare(Lit a, Lit b) {
			double sa;
			double sb;
			if (a.Sign == Assignment.True) sa = a.Var.PositiveRangeSize;
			else sa = a.Var.NegativeRangeSize;
			if (b.Sign == Assignment.True) sb = b.Var.PositiveRangeSize;
			else sb = b.Var.NegativeRangeSize;
			if (sa < sb) return -1;
			if (sa > sb) return 1;
			return 0;			
		}
		
	}
}

File: java/src/main/java/Reasoner/CNSAT/Decider.java
//
// Translated by CS2J (http://www.cs2j.com): 12.11.2016 09:40:46
//

package Reasoner.CNSAT;

import java.util.Comparator;
import java.util.LinkedList;
import java.util.List;
import java.util.Random;

public class Decider   
{
    public Decider() throws Exception {
    }

    public static Var decideRangeBased(List<Var> variables, CNSat solver) throws Exception {
        List<Lit> choices = new LinkedList<>();
        int vars = variables.size();
        for (int i = 0;i < vars;i++)
        {
            if (variables.get(i).getAssignment() != Assignment.Unassigned)
                continue;
             
            boolean hT = false;
            boolean hF = false;
            for (int j = 0;j < variables.get(i).getWatchList().size();j++)
            {
                if (!variables.get(i).getWatchList().get(j).getClause().getSatisfied())
                {
                    if (variables.get(i).getWatchList().get(j).getLit().getSign() == Assignment.True)
                    {
                        if (!hT)
                            choices.add(variables.get(i).getWatchList().get(j).getLit());
                         
                        hT = true;
                    }
                    else
                    {
                        if (!hF)
                            choices.add(variables.get(i).getWatchList().get(j).getLit());
                         
                        hF = true;
                    } 
                }
                 
                if (hT && hF)
                    break;
                 
            }
        }
        if (choices.size() == 0)
            return null;
         
        choices.sort(new Comparator<Lit>()
        {
            @Override
            public int compare(final Lit a, final Lit b)
            {
                double sa;
                double sb;
                if (a.getSign() == Assignment.True)
                    sa = a.getVar().getPositiveRangeSize();
                else
                    sa = a.getVar().getNegativeRangeSize();
                if (b.getSign() == Assignment.True)
                    sb = b.getVar().getPositiveRangeSize();
                else
                    sb = b.getVar().getNegativeRangeSize();
                if (sa < sb)
                    return -1;

                if (sa > sb)
                    return 1;

                return 0;
            }
        });

        Var v;
        Assignment ass = Assignment.False;
        //Uniform:
        //int idx = solver.Rand.Next(choices.Count);
        //smallest range:
        //int idx = 0;
        //largest range:
        int idx = choices.size() - 1;

        v = choices.get(idx).getVar();
        ass = choices.get(idx).getSign();
        DecisionLevel d = new DecisionLevel(solver.getDecisions().size());
        solver.getDecisionLevel().add(d);
        v.setAssignment(ass);
        solver.getDecisions().add(v);
        v.setReason(null);
        v.setDecisionLevel(d);
        return v;
    }

    public static Var decideActivityBased(List<Var> variables, CNSat solver) throws Exception {
        int vars = variables.size();
        Random r = solver.getRand();
        int init = r.nextInt(vars);
        Var v = null, next = null;
        int maxActivity = 0;
        //Search Lit with highest Activity
        if (solver.getCNSMTGSolver() == null)
        {
            for (int i = 0;i < vars;i++)
            {
                int p = (init + i) % vars;
                if (p < 0 || p >= vars)
                    System.out.println("p = " + p);
                 
                v = variables.get(p);
                if (v.getAssignment() == Assignment.Unassigned)
                {
                    if (maxActivity <= v.getActivity())
                    {
                        maxActivity = v.getActivity();
                        next = v;
                    }
                     
                }
                 
            }
            //Decide it
            if (next != null)
            {
                DecisionLevel d = new DecisionLevel(solver.getDecisions().size());
                solver.getDecisionLevel().add(d);
                double rel = next.getPositiveAppearance() + next.getNegativeAppearance();
                if (rel != 0)
                    rel = ((double)next.getPositiveAppearance()) / rel;
                else
                    rel = 0.5; 
                next.setAssignment((r.nextDouble() < rel) ? Assignment.True : Assignment.False);
                solver.getDecisions().add(next);
                next.setReason(null);
                next.setDecisionLevel(d);
                return next;
            }
             
        }
        else
        {
            init = r.nextInt(solver.getClauses().size());
            for (int i = 0;i < solver.getClauses().size();i++)
            {
                Clause c = solver.getClauses().get((i + init) % solver.getClauses().size());
                if (!c.getSatisfied() && c.getLiterals().size() > 1)
                {
                    if (!c.watcher[0].getLit().satisfied())
                    {
                        next = c.watcher[0].getLit().getVar();
                        DecisionLevel d = new DecisionLevel(solver.getDecisions().size());
                        solver.getDecisionLevel().add(d);
                        next.setAssignment(c.watcher[0].getLit().getSign());
                        solver.getDecisions().add(next);
                        next.setReason(null);
                        next.setDecisionLevel(d);
                        return next;
                    }
                    else if (!c.watcher[1].getLit().satisfied())
                    {
                        DecisionLevel d = new DecisionLevel(solver.getDecisions().size());
                        solver.getDecisionLevel().add(d);
                        next = c.watcher[1].getLit().getVar();
                        next.setAssignment(c.watcher[1].getLit().getSign());
                        solver.getDecisions().add(next);
                        next.setReason(null);
                        next.setDecisionLevel(d);
                        return next;
                    }
                    else
                        System.out.println("This shoud Never Happen!!");
                }
                 
            }
        } 
        return null;
    }

    public static Var decideVariableCountBased(List<Var> variables, CNSat solver) throws Exception {
        int vars = variables.size();
        Lit l = null;
        int maxCount = Integer.MAX_VALUE;
        int minCount = -1;
        for (int i = 0;i < vars;i++)
        {
            if (variables.get(i).getAssignment() != Assignment.Unassigned)
                continue;
             
            for (int j = 0;j < variables.get(i).getWatchList().size();j++)
            {
                if (!variables.get(i).getWatchList().get(j).getClause().getSatisfied())
                {
                    /*if(maxCount>variables.get(i).WatchList.get(j).Lit.VariableCount) {
                    							l = variables.get(i).WatchList.get(j).Lit;
                    							maxCount = l.VariableCount;
                    						}*/
                    if (minCount < variables.get(i).getWatchList().get(j).getLit().getVariableCount())
                    {
                        l = variables.get(i).getWatchList().get(j).getLit();
                        minCount = l.getVariableCount();
                    }
                     
                }
                 
            }
        }
        if (l == null)
            return null;
         
        Assignment ass = Assignment.False;
        Var v = l.getVar();
        ass = l.getSign();
        DecisionLevel d = new DecisionLevel(solver.getDecisions().size());
        solver.getDecisionLevel().add(d);
        v.setAssignment(ass);
        solver.getDecisions().add(v);
        v.setReason(null);
        v.setDecisionLevel(d);
        return v;
    }

}


